include/deal.II/fe/mapping_q.h

```diff
@@ -53,10 +132,805 @@ class MappingQ : public MappingQGeneric<dim, spacedim>
+   * interpolation, or if dim<spacedim, it asks the underlying manifold for
+   * the locations of interior points.
+   */
+  virtual std::vector<Point<spacedim>>
+  compute_mapping_support_points(
+    const typename Triangulation<dim, spacedim>::cell_iterator &cell) const;
+
+  /**
+   * Transform the point @p p on the real cell to the corresponding point on
+   * the unit cell @p cell by a Newton iteration.
+   */
+  Point<dim>
+  transform_real_to_unit_cell_internal(
+    const typename Triangulation<dim, spacedim>::cell_iterator &cell,
+    const Point<spacedim> &                                     p,
+    const Point<dim> &initial_p_unit) const;
+
+  /**
+   * Append the support points of all shape functions located on bounding
+   * lines of the given cell to the vector @p a. Points located on the
+   * vertices of a line are not included.
+   *
+   * This function uses the underlying manifold object of the line (or, if
+   * none is set, of the cell) for the location of the requested points. This
+   * function is usually called by compute_mapping_support_points() function.
+   *
+   * This function is made virtual in order to allow derived classes to choose
+   * shape function support points differently than the present class, which
+   * chooses the points as interpolation points on the boundary.
+   */
+  virtual void
+  add_line_support_points(
+    const typename Triangulation<dim, spacedim>::cell_iterator &cell,
+    std::vector<Point<spacedim>> &                              a) const;
+
+  /**
+   * Append the support points of all shape functions located on bounding
+   * faces (quads in 3d) of the given cell to the vector @p a. This function
+   * is only defined for <tt>dim=3</tt>. Points located on the vertices or
+   * lines of a quad are not included.
+   *
+   * This function uses the underlying manifold object of the quad (or, if
+   * none is set, of the cell) for the location of the requested points. This
+   * function is usually called by compute_mapping_support_points().
+   *
+   * This function is made virtual in order to allow derived classes to choose
+   * shape function support points differently than the present class, which
+   * chooses the points as interpolation points on the boundary.
+   */
+  virtual void
+  add_quad_support_points(
+    const typename Triangulation<dim, spacedim>::cell_iterator &cell,
+    std::vector<Point<spacedim>> &                              a) const;
+
+  // Make MappingQCache a friend since it needs to call the
+  // compute_mapping_support_points() function.
+  template <int, int>
+  friend class MappingQCache;
 };
 
+
+
+/**
+ * A class that implements a polynomial mapping $Q_p$ of degree $p$ on all
+ * cells. This class is completely equivalent to the MappingQ class and there
+ * for backward compatibility.
+ */
+template <int dim, int spacedim = dim>
+using MappingQGeneric = MappingQ<dim, spacedim>;
```

Comment:
@peterrum I have the `using` statement in `mapping_q.h` for the reason that I do not want to force users to change include files. If I had placed it in `mapping_q_generic.h`, a code with `#include <mapping_q.h>` and then using `MappingQGeneric` would generate an error now, but it wouldn't before. Therefore, we need it this way.

And by the way, I would not set the `early deprecate` warning on `MappingQGeneric` at this point because the class is so widely used that it would basically warn on >30% of all deal.II codes.

Reply:
> I have the using statement in mapping_q.h for the reason that I do not want to force users to change include files. If I had placed it in mapping_q_generic.h, a code with #include <mapping_q.h> and then using MappingQGeneric would generate an error now, but it wouldn't before. Therefore, we need it this way.

Makes sense. 

> And by the way, I would not set the early deprecate warning on MappingQGeneric at this point because the class is so widely used that it would basically warn on >30% of all deal.II codes.

I was tempted to suggest this ;)